internal/globalratelimit/runner/runner.go

```diff
@@ -36,62 +57,123 @@ func New(cfg *Config) *Runner {
 // Start starts the infrastructure runner
 func (r *Runner) Start(ctx context.Context) error {
 	r.Logger = r.Logger.WithValues("runner", r.Name())
+
+	// Set up the gRPC server for ratelimit xDS Config.
+	r.grpc = grpc.NewServer()
+
+	r.cache = cachev3.NewSnapshotCache(false, cachev3.IDHash{}, cache.NewLogrWrapper(r.Logger))
+
+	// Register xDS Config server.
+	cb := &test.Callbacks{}
+	discoveryv3.RegisterAggregatedDiscoveryServiceServer(r.grpc, serverv3.NewServer(ctx, r.cache, cb))
+
+	// Start and listen xDS gRPC config Server.
+	go r.serverXdsConfigServer(ctx)
+
+	// Start message Subscription.
 	go r.subscribeAndTranslate(ctx)
+
 	r.Logger.Info("started")
 	return nil
 }
 
-func (r *Runner) subscribeAndTranslate(ctx context.Context) {
-	// Subscribe
-	xdsIRCh := r.XdsIR.Subscribe(ctx)
-	for ctx.Err() == nil {
-		var xdsIRs []*ir.Xds
-		snapshot := <-xdsIRCh
-		r.Logger.Info("received a notification")
-		// Skip translation if state is empty
-		if len(snapshot.State) == 0 {
-			continue
-		}
+func (r *Runner) serverXdsConfigServer(ctx context.Context) {
+	addr := net.JoinHostPort(XdsGrpcSotwConfigServerAddress, strconv.Itoa(ratelimit.XdsGrpcSotwConfigServerPort))
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		r.Logger.Error(err, "failed to listen on address", "address", addr)
+		return
+	}
+	err = r.grpc.Serve(l)
+	if err != nil {
+		r.Logger.Error(err, "failed to start grpc based xds config server")
+	}
 
-		for _, value := range snapshot.State {
-			xdsIRs = append(xdsIRs, value)
-		}
+	<-ctx.Done()
+	r.Logger.Info("grpc config server shutting down")
+	r.grpc.Stop()
+}
+
+func (r *Runner) subscribeAndTranslate(ctx context.Context) {
+	// Subscribe to resources.
+	message.HandleSubscription(r.XdsIR.Subscribe(ctx),
+		func(update message.Update[string, *ir.Xds]) {
+			r.Logger.Info("received a notification")
 
-		// Translate to ratelimit infra IR
-		result, err := r.translate(xdsIRs)
-		if err != nil {
-			r.Logger.Error(err, "failed to translate xds ir")
-		} else {
-			if result == nil {
-				r.RateLimitInfraIR.Delete(r.Name())
+			if update.Delete {
+				err := r.addNewSnapshot(ctx, nil)
+				if err != nil {
+					r.Logger.Error(err, "failed to update the config snapshot")
+				}
 			} else {
-				// Publish ratelimit infra IR
-				r.RateLimitInfraIR.Store(r.Name(), result)
+				// Translate to ratelimit infra IR and ratelimit xDS Config.
+				rlIR, rvt, err := r.translate(update.Value)
+				if err != nil {
+					r.Logger.Error(err, "failed to translate xds ir and config")
+				} else {
```

Comment:
From line 113 to 131, can you do some opts on them ? Too many if else and make it less readable.

Reply:
wrapped config cache updating logic into a method, should be better now.